File: src/kairos/detectors/long_short_ratio.py
"""Long/Short Ratio anomaly detector using CoinGlass data.

Polls CoinGlass for per-symbol long/short ratio snapshots and emits anomalies
when the ratio reaches extreme levels (Z-score or absolute threshold) or
when it shifts rapidly.
"""

from __future__ import annotations

import logging
import threading
import time
from typing import Any

from kairos.detectors.base import AnomalyEvent, BaseDetector
from kairos.utils.zscore import ZScoreTracker

logger = logging.getLogger(__name__)


class LongShortRatioDetector(BaseDetector):
    """Detects anomalous long/short ratio levels and shifts.

    Two trigger modes (any match → anomaly):
      1. Absolute: long_rate exceeds abs_threshold (e.g. >80% or <20%)
      2. Z-score:  abs(z_score(long_rate)) > zscore_threshold
      3. Velocity: ratio shifted by more than velocity_threshold_pct within one poll interval

    Poll periodically via ``on_ls_snapshot(symbol, timestamp, long_rate, short_rate)``.
    """

    def __init__(self, config: dict) -> None:
        super().__init__(config)
        ls = config.get("longShortRatio", config)

        self.enabled = bool(ls.get("enabled", True))
        self.poll_interval_s = float(ls.get("pollIntervalSeconds", 300))
        self.abs_threshold = float(ls.get("absThreshold", 80.0))  # long_rate %
        self.zscore_threshold = float(ls.get("zscoreThreshold", 2.5))
        self.zscore_window = int(ls.get("zscoreWindow", 48))  # ~4h at 5min polls
        self.velocity_threshold_pct = float(ls.get("velocityThresholdPct", 15.0))  # % change
        self.min_notify_seconds = float(
            _parse_seconds_config(ls.get("minNotifyInterval", "30m"), 1800)
        )

        # symbol -> ZScoreTracker for long_rate %
        self._zscore: dict[str, ZScoreTracker] = {}
        # symbol -> last (timestamp, long_rate, short_rate)
        self._last: dict[str, tuple[float, float, float]] = {}
        # symbol_event_type -> cooldown timestamp
        self._last_notify: dict[str, float] = {}
        self._lock = threading.RLock()

    def on_ls_snapshot(
        self,
        symbol: str,
        timestamp: float,
        long_rate: float,
        short_rate: float,
    ) -> None:
        """Process one long/short ratio snapshot from polling."""
        if not self.enabled:
            return
        if long_rate is None or long_rate < 0 or long_rate > 100:
            return

        with self._lock:
            if symbol not in self._zscore:
                self._zscore[symbol] = ZScoreTracker(window=self.zscore_window, min_samples=5)

            zt = self._zscore[symbol]
            prev = self._last.get(symbol)

            # 1. Absolute threshold
            if long_rate >= self.abs_threshold or long_rate <= (100.0 - self.abs_threshold):
                label = "long_excessive" if long_rate >= self.abs_threshold else "short_excessive"
                if self._can_notify(symbol, label, timestamp):
                    self._emit_anomaly(symbol, timestamp, long_rate, short_rate, label, long_rate, None)
                    # still record for zscore but skip velocity if we just emitted
                    zt.add(long_rate)
                    self._last[symbol] = (timestamp, long_rate, short_rate)
                    return

            # 2. Velocity check (rapid shift vs previous poll)
            if prev is not None:
                prev_long, prev_short = prev[1], prev[2]
                if prev_long > 0:
                    shift_pct = abs(long_rate - prev_long) / prev_long * 100
                    if shift_pct >= self.velocity_threshold_pct:
                        label = "ls_velocity"
                        if self._can_notify(symbol, label, timestamp):
                            self._emit_anomaly(symbol, timestamp, long_rate, short_rate, label, shift_pct, None)

            # 3. Z-score against own history
            zs = zt.add_and_score(long_rate)
            if zs is not None and abs(zs) >= self.zscore_threshold:
                label = "ls_zscore"
                if self._can_notify(symbol, label, timestamp):
                    self._emit_anomaly(symbol, timestamp, long_rate, short_rate, label, long_rate, zs)

            self._last[symbol] = (timestamp, long_rate, short_rate)

    def _emit_anomaly(
        self,
        symbol: str,
        timestamp: float,
        long_rate: float,
        short_rate: float,
        reason: str,
        value: float | None,
        zscore: float | None,
    ) -> None:
        abs_val = abs(value) if value is not None else 0.0
        if abs_val >= 80.0 or (zscore is not None and abs(zscore) >= 4.0):
            severity = "HIGH"
        elif abs_val >= 60.0 or (zscore is not None and abs(zscore) >= 3.0):
            severity = "MEDIUM"
        else:
            severity = "LOW"

        self._emit(
            AnomalyEvent(
                symbol=symbol,
                event_type="long_short_ratio",
                severity=severity,
                data={
                    "long_rate": round(long_rate, 2),
                    "short_rate": round(short_rate, 2),
                    "ls_ratio": round(long_rate / short_rate, 4) if short_rate > 0 else None,
                    "reason": reason,
                    "trigger_value": round(value, 4) if value is not None else None,
                    "zscore": round(zscore, 4) if zscore is not None else None,
                    "threshold_abs": self.abs_threshold,
                    "threshold_zscore": self.zscore_threshold,
                    "threshold_velocity_pct": self.velocity_threshold_pct,
                },
                timestamp=timestamp,
            )
        )

    def _can_notify(self, symbol: str, label: str, now: float) -> bool:
        key = f"{symbol}__ls__{label}"
        last = self._last_notify.get(key, 0.0)
        if now - last < self.min_notify_seconds:
            return False
        self._last_notify[key] = now
        return True

    def update_config(self, config: dict) -> None:
        with self._lock:
            super().update_config(config)
            updated = LongShortRatioDetector(config)
            self.enabled = updated.enabled
            self.poll_interval_s = updated.poll_interval_s
            self.abs_threshold = updated.abs_threshold
            self.zscore_threshold = updated.zscore_threshold
            self.zscore_window = updated.zscore_window
            self.velocity_threshold_pct = updated.velocity_threshold_pct
            self.min_notify_seconds = updated.min_notify_seconds


def _parse_seconds_config(value: Any, default: float) -> float:
    """Parse a duration value (int/float/str like '30m') to seconds."""
    if isinstance(value, (int, float)):
        return float(value)
    s = str(value).strip().lower()
    if s.endswith("s"):
        return float(s[:-1])
    if s.endswith("m"):
        return float(s[:-1]) * 60
    if s.endswith("h"):
        return float(s[:-1]) * 3600
    return default

File: go/internal/detector/longshortratio.go
package detector

import (
	"context"
	"log/slog"
	"math"
	"sync"
	"time"

	"github.com/ArchdevilForge/kairos/internal/types"
	"github.com/ArchdevilForge/kairos/internal/utils"
)

// LongShortRatioDetector monitors CoinGlass long/short ratio snapshots and
// emits anomalies on 3 trigger modes:
//  1. Absolute: long_rate ≥ absThreshold (80%) or ≤ (100 - absThreshold)
//  2. Velocity:  per-interval shift ≥ velocityThresholdPct (15%)
//  3. Z-score:   rolling deviation ≥ zscoreThreshold (2.5)
//
// Ported from src/kairos/detectors/long_short_ratio.py.
type LongShortRatioDetector struct {
	BaseDetector

	enabled              bool
	absThreshold         float64 // long_rate %, default 80
	zscoreThreshold      float64
	zscoreWindow         int
	velocityThresholdPct float64
	minNotifyInterval    time.Duration

	// symbol -> RollingZScore for long_rate %
	zscore map[string]*utils.RollingZScore
	zsMu   sync.RWMutex
	// symbol -> (timestamp, longRate, shortRate)
	last   map[string]lsSnapshot
	lastMu sync.RWMutex
}

type lsSnapshot struct {
	ts        float64
	longRate  float64
	shortRate float64
}

// NewLongShortRatioDetector creates a detector from config.
func NewLongShortRatioDetector(cfg types.LongShortRatioConfig) *LongShortRatioDetector {
	d := &LongShortRatioDetector{
		BaseDetector: BaseDetector{
			Logger:    slog.Default().With("detector", "long_short_ratio"),
			events:    make(chan types.AnomalyEvent, 64),
			Cooldowns: make(map[string]time.Time),
		},
		enabled:              cfg.Enabled,
		absThreshold:         cfg.AbsThreshold,
		zscoreThreshold:      cfg.ZscoreThreshold,
		zscoreWindow:         cfg.ZscoreWindow,
		velocityThresholdPct: cfg.VelocityThresholdPct,
		minNotifyInterval:    parseDuration(cfg.MinNotifyInterval),
		zscore:               make(map[string]*utils.RollingZScore),
		last:                 make(map[string]lsSnapshot),
	}
	if d.absThreshold <= 0 {
		d.absThreshold = 80.0
	}
	if d.zscoreThreshold <= 0 {
		d.zscoreThreshold = 2.5
	}
	if d.zscoreWindow <= 0 {
		d.zscoreWindow = 48
	}
	if d.velocityThresholdPct <= 0 {
		d.velocityThresholdPct = 15.0
	}
	if d.minNotifyInterval <= 0 {
		d.minNotifyInterval = 30 * time.Minute
	}
	return d
}

func (d *LongShortRatioDetector) Name() string { return "long_short_ratio" }

func (d *LongShortRatioDetector) OnTicker(_ context.Context, _ types.Ticker)                      {}
func (d *LongShortRatioDetector) OnMetricsUpdate(_ context.Context, _ string, _ float64, _ float64) {}
func (d *LongShortRatioDetector) OnLiquidationSnapshot(_ string, _, _, _ float64)                  {}

// OnLSSnapshot processes one long/short ratio snapshot.
func (d *LongShortRatioDetector) OnLSSnapshot(symbol string, longRate, shortRate float64) {
	if !d.enabled {
		return
	}
	if longRate < 0 || longRate > 100 {
		return
	}
	now := time.Now()
	ts := float64(now.UnixMilli()) / 1000

	// Ensure z-score tracker exists.
	d.zsMu.Lock()
	zt, ok := d.zscore[symbol]
	if !ok {
		zt = utils.NewRollingZScore(d.zscoreWindow)
		d.zscore[symbol] = zt
	}
	d.zsMu.Unlock()

	d.lastMu.RLock()
	prev, hasPrev := d.last[symbol]
	d.lastMu.RUnlock()

	// 1. Absolute threshold.
	absLow := 100.0 - d.absThreshold
	if longRate >= d.absThreshold || longRate <= absLow {
		label := "long_excessive"
		if longRate <= absLow {
			label = "short_excessive"
		}
		key := symbol + "__ls__" + label
		d.cdMu.Lock()
		if canNotify(d.Cooldowns, key, now, d.minNotifyInterval) {
			d.cdMu.Unlock()
			d.emitLS(symbol, ts, longRate, shortRate, label, longRate, 0, now)
		} else {
			d.cdMu.Unlock()
		}
		// Still record the value for z-score history (without scoring).
		zt.Add(longRate)
		d.lastMu.Lock()
		d.last[symbol] = lsSnapshot{ts: ts, longRate: longRate, shortRate: shortRate}
		d.lastMu.Unlock()
		return
	}

	// 2. Velocity check (rapid shift vs previous poll).
	if hasPrev && prev.longRate > 0 {
		shiftPct := math.Abs(longRate-prev.longRate) / prev.longRate * 100
		if shiftPct >= d.velocityThresholdPct {
			key := symbol + "__ls__ls_velocity"
			d.cdMu.Lock()
			if canNotify(d.Cooldowns, key, now, d.minNotifyInterval) {
				d.cdMu.Unlock()
				d.emitLS(symbol, ts, longRate, shortRate, "ls_velocity", shiftPct, 0, now)
			} else {
				d.cdMu.Unlock()
			}
		}
	}

	// 3. Z-score against own history.
	zs := zt.Add(longRate)
	if zs != 0 && d.zscoreThreshold > 0 && math.Abs(zs) >= d.zscoreThreshold {
		key := symbol + "__ls__ls_zscore"
		d.cdMu.Lock()
		if canNotify(d.Cooldowns, key, now, d.minNotifyInterval) {
			d.cdMu.Unlock()
			d.emitLS(symbol, ts, longRate, shortRate, "ls_zscore", longRate, zs, now)
		} else {
			d.cdMu.Unlock()
		}
	}

	d.lastMu.Lock()
	d.last[symbol] = lsSnapshot{ts: ts, longRate: longRate, shortRate: shortRate}
	d.lastMu.Unlock()
}

func (d *LongShortRatioDetector) Events() <-chan types.AnomalyEvent { return d.events }

func (d *LongShortRatioDetector) Reset() {
	d.zsMu.Lock()
	clear(d.zscore)
	d.zsMu.Unlock()
	d.lastMu.Lock()
	clear(d.last)
	d.lastMu.Unlock()
	d.cdMu.Lock()
	clear(d.Cooldowns)
	d.cdMu.Unlock()
}

// UpdateConfig hot-reloads configuration at runtime.
func (d *LongShortRatioDetector) UpdateConfig(cfg types.LongShortRatioConfig) {
	d.BaseDetector.mu.Lock()
	defer d.BaseDetector.mu.Unlock()
	d.enabled = cfg.Enabled
	if cfg.AbsThreshold > 0 {
		d.absThreshold = cfg.AbsThreshold
	}
	if cfg.ZscoreThreshold > 0 {
		d.zscoreThreshold = cfg.ZscoreThreshold
	}
	if cfg.ZscoreWindow > 0 {
		d.zscoreWindow = cfg.ZscoreWindow
	}
	if cfg.VelocityThresholdPct > 0 {
		d.velocityThresholdPct = cfg.VelocityThresholdPct
	}
	if cfg.MinNotifyInterval != "" {
		d.minNotifyInterval = parseDuration(cfg.MinNotifyInterval)
	}
}

func (d *LongShortRatioDetector) emitLS(symbol string, ts, longRate, shortRate float64,
	reason string, value, zs float64, nowTime time.Time) {

	absVal := math.Abs(value)
	var severity types.Severity
	if absVal >= 80.0 || (zs != 0 && math.Abs(zs) >= 4.0) {
		severity = types.SeverityHigh
	} else if absVal >= 60.0 || (zs != 0 && math.Abs(zs) >= 3.0) {
		severity = types.SeverityMedium
	} else {
		severity = types.SeverityLow
	}

	lsRatio := 0.0
	if shortRate > 0 {
		lsRatio = math.Round(longRate/shortRate*10000) / 10000
	}

	data := map[string]any{
		"long_rate":             math.Round(longRate*100) / 100,
		"short_rate":            math.Round(shortRate*100) / 100,
		"ls_ratio":              lsRatio,
		"reason":                reason,
		"trigger_value":         math.Round(value*10000) / 10000,
		"zscore":                math.Round(zs*10000) / 10000,
		"threshold_abs":         d.absThreshold,
		"threshold_zscore":      d.zscoreThreshold,
		"threshold_velocity_pct": d.velocityThresholdPct,
	}
	if zs == 0 {
		data["zscore"] = nil
	}

	evt := NewEvent(symbol, "long_short_ratio", string(severity), data)
	evt.Timestamp = ts
	select {
	case d.events <- evt:
	default:
		d.Logger.Warn("long_short_ratio event channel full, dropping event")
	}
}
